Docs4Doge : create "getting started" section (#2568)
dogecoin/dogecoin: doc/getting-started.md

This tutorial will help you to go through the basics to use them !

> **Warning :** For simplicity, explained commands assume executables are under `PATH` environment variable.  
> If needed, specify their location by doing `PATH=$PATH:/path/to/executables` or use their path as command like :
> ```bash
> $ /path/to/dogecoin-cli [arguments ...]
> ```

### Table of contents

1. [Run a dogecoin node](#run-a-dogecoin-node)
2. [Introduction to JSON-RPC API](#introduction-to-json-rpc-api)
    * [Create a wallet](#create-a-wallet)
    * [Send a transaction](#send-a-transaction)
    * [Inspect blocks and transactions](#inspect-blocks-and-transactions)
3. [Node configuration](#node-configuration)
    * [Mainnet, testnet and regnet](#mainnet-testnet-and-regnet)
    * [RPC credentials](#rpc-credentials)
    * [Ports](#ports)
    * [Memory](#memory)
    * [Mode debug](#mode-debug)

## Run a dogecoin node

To launch your node, you can choose `dogecoind` to run an headless server or `dogecoin-qt` for a server with a GUI.

In a terminal, run :
```bash
# Headless node

> $ dogecoind -printtoconsole

Reviewer: why `-printtoconsole` and not `-daemon`?

Author: Personal experience where I was confused about not having any output on the first time I runned a node. I find it more intuitive to display first some output and then introduce debug.log.

Why `-daemon` and not `&` ?

Reviewer: because `-daemon` detaches from the user session and has the root process as its parent, whereas `&` has the terminal process as its parent. Meaning that if the terminal gets killed, dogecoind gets killed. What if the terminal gets killed because it ran OOM due to all the output you fed it with `-printtoconsole`?

Author: Don't you think it may be good to keep `-printtoconsole` for intro here, and use daemon in the example for configuration later ? Like it they will have both, with a first one showing some output, and finally a way to daemonize their node ?

Author: Changed to use `-daemon`. Let me know if it's ok with the current modification.

Reviewer: Yes. Looks good!

# GUI
$ dogecoin-qt

> ```

Author: It's because commands are listed on the previous sentence, you would remove it ? I think it's great to keep a block of code here to make it more obvious.

Otherwise you're suggesting 2 blocks with a small sentence before each of them ?

> ```

Reviewer: > Otherwise you're suggesting 2 blocks with a small sentence before each of them ?

Yeah. Now it's a code block with comments - it looks odd and confusing.

> ```

Author: Updated, tell me if it's ok for you.

> ```

Reviewer: It's still the same for me?

> ```

Author: My bad, I had merge conflict.

> ```

Reviewer: Yes, this is better!

> *Use `--help` to see all available options for each executables.*

Author: ok

> **Warning :** The rest of the guide will assume the use of an headless node. The RPC server is not exposed with `dogecoin-qt` until you activate the `-server` option.

## Introduction to JSON-RPC API

Dogecoin Core expose a JSON-RPC interface. With it, you can request information about blocks and transactions, send transactions to the networks, manage your wallet and so on...

To interact with the interface, it relies on HTTP requests. Using urls like `http://user:password@127.0.0.1:22555/` and sending commands through json format is feasible.

Some tools like libraries can be built on top of it. `dogecoin-cli` is the default shell client provided with Dogecoin Core to use this API.

The RPC API documentation is available from [bitcoin](https://developer.bitcoin.org/reference/rpc/).  
To have an overview of all commands, see also :

```bash
# List all commands
$ dogecoin-cli help

# Get help for a specific command
$ dogecoin-cli help COMMAND
```
> **Tips** : Abuse of the help menus is allowed ! They can answer to multiple questions.

### Create a wallet

> 

Author: May be interesting to add `importprivkey`

Reviewer 2: 1.14.4 added a GUI for that, just saying....

Author: ? Doing this through the CLI is useful too

Author: But it's the command used to do this, the problem is more the way to use it. I totally agree with the security risk, but I suppose not showing it won't solve the problem.

In a scenario where someone is playing with the CLI and want to import some keys, he will google it and find this info. Explained or not, these risks exists. Don't you think it's more interesting to show the command and explain briefly associated risks ?

Also, devs may need this information without being exposed to real risk in a development environment.

Reviewer: > Don't you think it's more interesting to show the command and explain briefly associated risks ?

It depends on the intended audience for the document. If I look at it from a getting started perspective, then I'd leave it out and instead just note that it's important to make wallet.dat backups. Perhaps it's an idea to implement sweeping of private keys rather than import, and then list that instead of importing them as a best practice? I.e. leave it out for now, build sweep into 1.14.6's RPC, document that, then build it into the UI under Operation Such Frenzy and we're Gucci?

Author: Agree to keep it has it, it's a good starting point and this information is not hard to find. But yeah, I was targeting more blockchain devs rather than basic Dogecoin users.

To receive or recover coins, you will need a wallet address and a private key. The *wallet address* is a shareable information to receive dogecoins from all over the world, while the *private key* is a sensitive information who give the ownership on coins associated to a public address.

By default, you already have a wallet address and a private key available.

> List it using `getaddressesbyaccount` :

Author: Use *"List wallet addresses"* instead of *"List it"*

```bash
$ dogecoin-cli getaddressesbyaccount ""
[
  "DBZc6qeBuo7g8La7L3Lhaf39MsJSgbzpMX"
]
```

Use `getnewaddress` to generate a new wallet address :
```bash
$ dogecoin-cli getnewaddress
DC13A8qUgAu5rTWweb44TwXZNQbxDYhyo2
```

And `dumpprivkey` to find the corresponding private key.
```bash
$ dogecoin-cli dumprivkey DC13A8qUgAu5rTWweb44TwXZNQbxDYhyo2
QSqPGRwxeqv8KNEU85jw128g7rt5ognjqaSvjmeouuFAjJ8Bq3ZR

> ```

Author: It depends on who is getting started. When I was doing it I've more developers in mind rather than common users. From this point of view, generating private keys, list them, import and dump them are basics commands to play with Dogecoin Core.

> ```

Reviewer: For developers, I would write a completely different guide though and have different guides for different development goals. For example, someone handling public coin should never use the wallet because it forces a solution to be custodial. Someone developing integration with their storefront maybe could use the wallet, but I'd still be extremely cautious with recommending that because you really don't want your keys exposed like that. A developer should learn how to use extended public keys so that they can segregate secrets from address generation - and maybe even keep their keys safely in a HW wallet.

Let's clarify this first because it impacts the entire document: devs is not the same audience as the "technical users" you quoted in your summary. You want this to be pedagogical and beginner friendly? Then it should teach the right things, safe things, and not dangerous things.

> ```

Author: `dumpprivkey` is a basic command, it's definitively needed to control your private key. It creates security risks for sure, but I feel it's more a security by ignorance and I really dislike it. "Not your keys, not your coins" => No access to your keys, no control of your coins.

People have to educate themselves about private/public key. This command is also coming from [Mastering Bitcoin](https://github.com/bitcoinbook/bitcoinbook/blob/develop/ch05.asciidoc).

> ```

Author: Also this get is definitively not targeting some kind of basic hodler. I suppose main Dogecoin usage are :
1. people Using exchanges
2. People using non-custodial
3. People running a node with their wallet, using GUI
4. People running a node with their wallet, using CLI.

Like it's really specific type of user who will come here, and when I think about Dogecoin core user it's more someone with programming skills already, people who will read it are filtered.

> ```

Reviewer: So from my perspective, this guide basically covers advanced users under 3 and 4 (3 because any command with `dogecoin-cli` can be done on the RPC console in QT) Maybe someone that wants to automate some personal coin sends... such as *whenever Elon tweets, send some Dogecoin to my fav exchange so I can sell at peak and buy back low later*.

> People have to educate themselves about private/public key. This command is also used in Mastering Bitcoin.

That section also mentions that there are HD wallets that have key derivation. That's what we have (basically right now we have what they call Type-1 HD wallets). For these, being careless with private keys can really hurt because of the way BIP32 derivation works. One key and some signatures can in theory expose the entire chain of keys. So this makes it much riskier to expose private keys like that.

> "Not your keys, not your coins" => No access to your keys, no control of your coins.

Owning your keys is not the same as exposing your keys. wallet.dat is a documented bdb container that can be read also outside of Dogecoin Core (in case you were worried about vendor lock-in on an open-source SW) I understand the educational value of what you're trying to achieve, but if you want to do this, then this should not be part of getting started, but part of a "how to secure your coin - let's try this on testnet" type of guide. `dumpprivkey` is really the worst possible way to securely manage your keys, because it removes all security and just outputs a secret in plaintext.

> ```

Author: Is there any other command to export your private key securely with `dogecoin-cli` ? I don't find any, and I'm wondering if encryption is a layer build on top of this command.

> ```

Reviewer: `backupwallet` doesn't decrypt keys.

> ```

Author: `dumprivkey` replaced by `backupwallet`.

> **Tips :** Dogecoin wallets addresses mainly start with the letter `D` and private keys with `Q`.

> You now have two wallet addresses to share to anyone who want to give you some dogecoins !

Author: Best practices (do you know from who it's coming from?) for anonymity recommend that, but it's definitely not how it's used.

I even see it as inconvenient, should I deal with tens of wallets ? And for people who convert their money and need to pay taxes, I suppose it would be a nightmare to behave like that. Anonymity is a real concern, but I won't be the guy who insist to never reuse a wallet for each transaction to be anonymous while everyone is giving IDs to exchanges.

We can suggest this practice, but I doubt about removing sentences related to reusing a wallet because it's in conflict of this "best practice".

> You now have two wallet addresses to share to anyone who want to give you some dogecoins !

Reviewer: It's been a best practice for years. See https://en.bitcoin.it/wiki/Address_reuse for an overview. If you don't believe me, please read the code in this repository. Also read [this](https://web.archive.org/web/20180712015726/http://www.nilsschneider.net/2013/01/28/recovering-bitcoin-private-keys.html)

It's not a best practice with quotes. It's an actual best practice that is implemented throughout the wallet, and warnings are generated when reusing address. And it's a security thing even more than a privacy thing. Please, I'm not bs-ing you.

> You now have two wallet addresses to share to anyone who want to give you some dogecoins !

Author: I already heard about this practice, I believe you and not thinking you're telling me any bs. Just it is from my pov definitively not how cryptos are used and it's even unpractical for many people, I see it more as an idealistic recommendation.

The security issue linked in your comment is also coming from another problem. Could have been prevented without reusing addresses for sure, but also could have been prevented by not having this random number generation issue. Interesting, but I'm not sure about what should I think of it.

It was probably a huge surprise for bitcoin devs to see people reusing their addresses !

> You now have two wallet addresses to share to anyone who want to give you some dogecoins !

Reviewer: > but also could have been prevented by not having this random number generation issue.

New weaknesses get found in algorithms all the time. Not reusing addresses not only segregates coin over multiple keys but also retains a forward protection because as long as there is just a hash of the public key known, an attacker not only needs to attack the signing algorithm, but also be able to brute-force or crack the hashed value into a pubkey. So by not reusing addresses, you're protecting yourself against future exploits. We're talking people's assets here. Not yours, not mine, but other people's. 

If we want to educate people, then we must educate the correct usage, not whatever people do that haven't given a single thought about the implications of their actions. Address reuse is bad - it really is that simple. Advocating for it in documentation on the Dogecoin Core repository goes against the code that is implemented here. I think it borders on neglect because we KNOW that it's bad, so let's not support it.

> You now have two wallet addresses to share to anyone who want to give you some dogecoins !

Author: Using the following :
> You now have two wallet addresses to share to anyone who want to give you some dogecoins ! Consider avoiding address reuse for anonymity and security reasons.

?

> You now have two wallet addresses to share to anyone who want to give you some dogecoins !

Reviewer: Yes, that can work. We can even link to the bitcoin wiki?

> You now have two wallet addresses to share to anyone who want to give you some dogecoins !

Author: Added with the link

### Send a transaction

> Dogecoin Core follow the [Unspent Transaction Output (UTXO)](https://en.wikipedia.org/wiki/Unspent_transaction_output) model to manage balances. To own some dogecoins is like having a list of unspent transactions. To spend them, you have to compose a new transaction with received unspent transaction.

Reviewer: - Dogecoin, not Dogecoin Core.
- Would change "unspent transactions" to "unspent outputs" because that's more concise.

> Four commands to execute to be able to tip : `listunspent`, `createrawtransaction`, `signrawtransaction` and `sendrawtransaction`.

Reviewer: Why not just `sendtoaddress`?

Author: Mainly because I followed mastering bitcoin, and they were using this flow. I don't know personally used this command, also I was concern about giving granularity to choose transaction/wallets.

I didn't explore the command enough, but I was also worried to have the same behavior I experienced on dogechain, who pick UTXO from multiple wallet and send money. It gives more the low-level information.

Reviewer: Do you need that to get started? Or is that the advanced course?

Author: Depend on who you are, again it's also targeting devs who may want to create application, and for this guy I supposed it's useful and a basic information.

Reviewer: What type of application?

Author: Whatever it is, in this case something which implement transactions. Let's say a video game for example.

Reviewer: My question is more what does it implement transactions for? What does it do? 

Does it pay out pre-funded DOGE to players? Then sure, but you can probably use `sendtoaddress` or worst case `fundrawtransaction` and suddenly you don't have to write your own coin selection algorithm, unless you think the one in the wallet is bad (but if you have an opinion about that, then you likely don't need to read this guide.) 

Does it allow players to do micro purchases / trades (think steam's marketplace)? Then you'll probably want payment channels (for speed) and your own non-custodial wallet software inside your game (to not become a money transmitter / bank) so you won't use the wallet for transactions.

I really don't see much of a use-case here except for wanting to create your own coin selection algorithm. But that's hardly "getting started" material. I also strongly doubt whether we should target developers in this guide as I mentioned above: the recommendations for developers to use Dogecoin Core to interface with the blockchain are much more involved and require much more thought than we're giving to this right now. I've been helping devs bootstrap crypto applications for over 7 years now and the things you'll need to run your application, even if it's just for accepting simple payments, are much greater than what we're describing here. 

I'm willing to work on another set of documents for developers with you if you aspire to cover that reader segment too, but if we really want to use Dogecoin Core for this, we'll have to do a bunch of enhancements to the actual software too. I do think it's a great goal to have though.

Author: I think we should consider targeting developer rather than simple users, at first. Because we may expect developers to work on user documentation, while the opposite isn't possible. In my mind, it lacks devs around the project so it's why I try to target potential developers, and in any case it uses CLI mainly, it reduces drastically potential readers.

You're definitively more skilled than me to know what is needed for devs to build an application, it's not enough, but I hope it's more resources than it currently is. I see it more as a starting point to build more complete documentation. In this guide, I'm translating "Mastering Bitcoin" content into the documentation. That's why you see the transaction process split, the transaction decoded in two steps. Because they are doing things in that way.

Author: It's good to add `sendtoaddress`, but I think it can be good to keep the process split to. For educational purpose it may be great, and it can give idea to people, where you can build your transaction and sign it later.

Showing commands split, and then show a transaction with `sendtoaddress`, to show there is a simple way to do it.

Reviewer: > In my mind, it lacks devs around the project so it's why I try to target potential developers

In that case, I propose to postpone this. Your description of this PR was to target users, not so much devs. Creating developer documentation is very different from "getting started" and should be much more about integration patterns and how to realize them. This is not what this document does at all. 

>  I'm translating "Mastering Bitcoin" content into the documentation

That's published under CC4 ShareAlike which is not compatible with MIT under which this document would ideally be published here. So now we may have an issue.

Author: It reuses the process, not the content. Nothing is copied, and used commands are done with new wallet/transaction/whatever. It shouldn't have any issue regarding copyright.

Difficult to be accurate about who it's targeting, the guy who will use Dogecoin through CLI is someone with programming skills, so a potential developer. Idk what is a typical "Dogecoin core" user.

I agree it would be great to have a well-delimited developer doc and a user doc, for now what I'm suggesting is a getting started guide for people with programming skills to be able to play with Dogecoin core, so simultaneously potential devs. It's to help people to jump in with the core to expect in the future to go in this direction and expect to get a more documented software.
+ potentially guys from the foundation were speaking about "dogepedia", it may complete this gap for more common users. Perhaps one day they will help on docs too.

Personally, I won't work on dev docs. First because I do not have the experience/skill + I would like to finish my open PRs and I will go away.

Reviewer: > Difficult to be accurate about who it's targeting

Okay, so let's do a simple `getbalance` and `sendtoaddress`, and the more advanced manual coin selection both? 

> I would like to finish my open PRs and I will go away.

😕 I guess that means I'll get to take over maintenance of the docs you're creating...

Author: The entire documentation in overall need to be maintained and to evolve, like you said more content is needed with a developer section with detailed instruction, API documentation, security recommendation, contribution guides, whatever. Idk what would come from bitcoin and be specific to Dogecoin, but there are many areas of improvement for the documentation, and in any case its endless work through the software lifecycle.

We reworked installation instruction, may add this getting started section + docker. The idea was to facilitate developer path when they start with Dogecoin Core, to help to involve more people inside the ecosystem. I found it rude to start when I first come here, I hope what we are doing is useful to make it easier.

I added content for `getbalance` and `sendtoaddress`.

##### listunspent

Find a list of UTXOs associated to any wallet addresses.

```bash
# Syntax
$ dogecoin-cli listunspent minconf maxconf '["address", ...]'

# Example

> $ dogecoin-cli listunspent 1 9999999 '["nnJDY1xCRgWQc7vBXHUPMPsEynuZW23Y3P"]'

Reviewer: This uses a testnet address, and that makes it confusing, suggest to just replace the occurrences of the testnet address, here and below, with an actual Dogecoin base58 address starting with `D`

Author: I saw it also as a way to introduce testnet.

```bash
# Find block hash from his height
$ dogecoin-cli getblockhash 69
3d2def20cd0d3aca148741ef469bda11647a3040d7669c82745d03c728706a8b

# Get block data
$ dogecoin-cli getblock 3d2def20cd0d3aca148741ef469bda11647a3040d7669c82745d03c728706a8b
{
  "hash": "3d2def20cd0d3aca148741ef469bda11647a3040d7669c82745d03c728706a8b",
  "confirmations": 7816,
  "strippedsize": 190,
  "size": 190,
  "weight": 760,
  "height": 69,
  "version": 1,
  "versionHex": "00000001",
  "merkleroot": "695ce4208fa7a87ef9e99805b0910dc129058ecdceb5cef7e25f71dcdc7936db",
  "tx": [
    "695ce4208fa7a87ef9e99805b0910dc129058ecdceb5cef7e25f71dcdc7936db"
  ],
  "time": 1386475225,
  "mediantime": 1386475209,
  "nonce": 3923708672,
  "bits": "1e0ffff0",
  "difficulty": 0.000244140625,
  "chainwork": "0000000000000000000000000000000000000000000000000000000004600460",
  "previousblockhash": "ffa69e04f928b84f19d84da25fb544340e54dca6c03c33930da245719e61c5ea",
  "nextblockhash": "44bf8abbbb96d4dcfb95df563e606c37987133ea3e013b23bbddde8d7f905fdd"
}
```

See the `tx` field to find the list of all block transactions. Only one exist in the block 69th, a coinbase transaction.

To get transaction details :

```bash

> $ dogecoin-cli getrawtransaction 695ce4208fa7a87ef9e99805b0910dc129058ecdceb5cef7e25f71dcdc7936db

Reviewer: You can add a boolean to getrawtransaction to get the json:

```
Arguments:
1. "txid"      (string, required) The transaction id
2. verbose       (bool, optional, default=false) If false, return a string, otherwise return a json object
```

PS: are you running with `-txindex`?

Author: I wasn't running with `-txindex`. You would like to see the json display ?

Reviewer: You're doing 2 commands for something that can be done in one... what does the hex encoded transaction add to someone getting started?

Author: Are you speaking about `gettransaction` ? Could you perform your desired command on testnet for `695ce4208fa7a87ef9e99805b0910dc129058ecdceb5cef7e25f71dcdc7936db` and share the output. My pi node crashed with `-txindex`, I'm not sure if I can run it today.

Reviewer: `gettransaction` is for wallet tx, `getrawtransaction` is for mempool tx, or if using `-txindex`, all mined+mempool tx.

Assuming you meant mainnet and not testnet, because there is no such tx on testnet.

Without `-txindex`:
```
$ dogecoin-cli getrawtransaction 695ce4208fa7a87ef9e99805b0910dc129058ecdceb5cef7e25f71dcdc7936db
error code: -5
error message:
No such mempool transaction. Use -txindex to enable blockchain transaction queries. Use gettransaction for wallet transactions.
```

With `-txindex`:
```
$ dogecoin-cli getrawtransaction 695ce4208fa7a87ef9e99805b0910dc129058ecdceb5cef7e25f71dcdc7936db
01000000010000000000000000000000000000000000000000000000000000000000000000ffffffff0e04d9eea3520101062f503253482fffffffff0100ac6156be23000023210340a42a5ad6c4c0cd5ae539657032e0a359bd3e0f95771f34d71691b13460a624ac00000000
```

With `-txindex` and `verbose=1` (last arg):
```
$ dogecoin-cli getrawtransaction 695ce4208fa7a87ef9e99805b0910dc129058ecdceb5cef7e25f71dcdc7936db 1
{
  "hex": "01000000010000000000000000000000000000000000000000000000000000000000000000ffffffff0e04d9eea3520101062f503253482fffffffff0100ac6156be23000023210340a42a5ad6c4c0cd5ae539657032e0a359bd3e0f95771f34d71691b13460a624ac00000000",
  "txid": "695ce4208fa7a87ef9e99805b0910dc129058ecdceb5cef7e25f71dcdc7936db",
  "hash": "695ce4208fa7a87ef9e99805b0910dc129058ecdceb5cef7e25f71dcdc7936db",
  "size": 109,
  "vsize": 109,
  "version": 1,
  "locktime": 0,
  "vin": [
    {
      "coinbase": "04d9eea3520101062f503253482f",
      "sequence": 4294967295
    }
  ],
  "vout": [
    {
      "value": 393004.00000000,
      "n": 0,
      "scriptPubKey": {
        "asm": "0340a42a5ad6c4c0cd5ae539657032e0a359bd3e0f95771f34d71691b13460a624 OP_CHECKSIG",
        "hex": "210340a42a5ad6c4c0cd5ae539657032e0a359bd3e0f95771f34d71691b13460a624ac",
        "reqSigs": 1,
        "type": "pubkey",
        "addresses": [
          "D5TjotnkoXekNQBCn54CBWxjEzCJMMe7WS"
        ]
      }
    }
  ],
  "blockhash": "3d2def20cd0d3aca148741ef469bda11647a3040d7669c82745d03c728706a8b",
  "confirmations": 3964556,
  "time": 1386475225,
  "blocktime": 1386475225
}
```

Author: Thanks, updated.

Author: Yeah probably, I think I used it and I missed adding this information. If you query you own transaction, would you need `txindex` ?

We may change the introduction sentence to explain it, and suggest `txindex` if user want to query other txs.

Author: I added a warning about txindex at the top of the section, explaining to use this option if the query is not related to user wallet.

{
  "txid": "695ce4208fa7a87ef9e99805b0910dc129058ecdceb5cef7e25f71dcdc7936db",
  "hash": "695ce4208fa7a87ef9e99805b0910dc129058ecdceb5cef7e25f71dcdc7936db",
  "size": 109,
  "vsize": 109,
  "version": 1,
  "locktime": 0,
  "vin": [
    {
      "coinbase": "04d9eea3520101062f503253482f",
      "sequence": 4294967295
    }
  ],
  "vout": [
    {
      "value": 393004.00000000,
      "n": 0,
      "scriptPubKey": {
        "asm": "0340a42a5ad6c4c0cd5ae539657032e0a359bd3e0f95771f34d71691b13460a624 OP_CHECKSIG",
        "hex": "210340a42a5ad6c4c0cd5ae539657032e0a359bd3e0f95771f34d71691b13460a624ac",
        "reqSigs": 1,
        "type": "pubkey",
        "addresses": [
          "D5TjotnkoXekNQBCn54CBWxjEzCJMMe7WS"
        ]
      }
    }
  ]
}
```

`vin` and `vout` will give you information about transaction inputs and outputs.

## Node configuration

A server contain a lot of parameter to configure, from memory limits, to RPC credentials, ports... There is two ways to prepare your perfect node !

Using directly `dogecoind` options, `--help` will display all configuration parameters.

**Command example :**
```bash

> $ dogecoind -testnet -datadir=/path/to/dir -rpcuser=Shibetoshi -rpcpassword=Nakamoto

Reviewer: would not add `-testnet` here, and also note that `-rpcuser` and `-rpcpassword` are not the recommended way to run a local node, but instead there is the auth cookie mechanism. Changing the datadir is dangerous too, so perhaps we can use a different example?

Author: > Changing the datadir is dangerous too.

I understand for credentials (related to bash history issues), but what are the risks for changing datadir ?

Why not adding `-testnet` here ? It's just to show a node with some configurations, testnet may be common, even more for devs.

Author: So it's not dangerous but time-consuming. Couldn't we rely on logic of users ? It's showing possibilities, and at worst it's something they can stop if they already synced the blockchain.

Reviewer: Well, if you sent yourself coin to the wallet in datadir A and then you create a new wallet in datadir B, you better not overwrite those wallet.dat files from B to A...

It complicates things and shouldn't be in a beginner's guide imho. I guess you really feel this is important though? Why?

Author: I think it's important to be able to manage the directory where data are stored, whichever is the program.

And they have to modify it manually to make it works, `/path/to/dir` location will break the command.

Reviewer: Alright. To be practical then: if you really insist on keeping it, could you please make a section about this and not introduce it in the example of passing arguments? Please also document the impact on using dogecoin-cli when having a custom datadir and/or rpc credentials there.

As for passing arguments, maybe we can make a case around `-paytxfee` and `-maxconnections` instead? Those are easier to comprehend too and don't impact other commands.

Author: Okay having configuration example of `dogecoind` & the configuration file with the following :

```
dogecoind -daemon -server -listen -paytxfee=0.01 -sendfreetransactions=1 -maxconnections=150
```
```
daemon=1
server=1
listen=1
paytxfee=0.01
sendfreetransactions=1
maxconnections=150
```
I'm not sure if `server` and `listen` are useful with `dogecoind`, seem default settings. Possibly you suggested [this config](https://github.com/dogecoin/dogecoin/pull/2568#discussion_r740535233) for `dogecoin-qt`.

Reviewer: yeah you can remove the `server` and `listen`, keeps it shorter.

Author: Example replaced.

```

Otherwise, a node is configurable through a `dogecoin.conf` file. Create it in the directory defined with the `datadir` setting, `$HOME/.dogecoin` by default, or specify the file location with `-conf`.

**dogecoin.conf example :**
```

> testnet=1
> datadir=/path/to/directory
> rpcuser=Shibetoshi
> rpcpassword=Nakamoto

Reviewer: Maybe something more common, like:

```suggestion
daemon=1
server=1
listen=1
sendfreetransactions=1
maxconnections=150
```

Author: It may be interesting to keep the same configuration between this config file example and the previous command where settings are passed through cli, it was the idea.

It would prefer to keep an example with testnet and datadir, but if you want to only go with differents settings, I don't care.

```
You can see a more concrete example [here](/contrib/debian/examples/dogecoin.conf).

### Mainnet, testnet and regnet

To perform tests, for example to create a blockchain based application, you may need a sandbox environment. Multiple networks exists for this purpose.

**Mainnet** : The main network where real transaction operate.  
**Testnet** : The test network, with peers.  
**Regnet** : The regression network, to test without peers and create blocks on-demand.

*Mainnet* is the network used by default. To enable *testnet*, with `dogecoind` use the `-testnet` option and `testnet=1` as `dogecoin.conf` field.

To enable *regnet*, use `-regtest` option.

> **Tips :** Think to specify the network when you use `dogecoin-cli`.

### RPC credentials

A username and a password is needed to control the RPC interface. When no credentials are provided, dogecoin use a [special cookie](https://bitcoin.org/en/release/v0.12.0#rpc-random-cookie-rpc-authentication) to randomly generate it.

> For security reasons, because the RPC server may process sensitive data like private keys, you can define your own identifiers using `rpcuser` and `rpcpassword` settings.

Reviewer: The cookie is generally more secure than the credentials though, because it's randomized on startup?

Instead, I'd recommend to UNDER NO CIRCUMSTANCE open the RPC port to the public internet.

Author: Changed by the following :
> A username and a password is needed to control the RPC interface. When no credentials are provided, dogecoin use a special cookie to randomly generate it. It's possible to define your own identifiers using rpcuser and rpcpassword settings.

I removed the misleading sentence about security.

Reviewer: For now, sure, we catch grammar in round 2 ❤️

### Ports

A node exposes up to 2 ports. One mandatory port for the **Peer to Peer Network (P2P)** to communicate with other nodes, a second optional port expose the RPC API.

| Function | mainnet | testnet | regtest |
| :------- | ------: | ------: | ------: |
| P2P      |   22556 |   44556 |   18444 |
| RPC      |   22555 |   44555 |   18332 |

To configure them use `port` and `rpcport` settings.

### Memory

Dogecoin Core use important ressources for some environment, it may be interesting to restrict them.

Multiple parameter to play with, listed here with their default values :
```bash
$ dogecoind -dbcache=450 -maxmempool=300 -maxconnections=125
```
- **dbcache :** UTXO database cache size.
- **maxmempool :** Limit memory pool size.
- **maxconnections :** Maximum peer connections.

> 

Reviewer: I think that none of this helps with getting started.

Author: It can be useful to deal with memory configuration when you're starting your node no ?

Reviewer: I'm not sure. changing `maxmempool` can have effects to what fees your node is willing to relay. Is that something for beginners? `dbcache` - I'd only tune that in combination with debug=bench and some trial and error, which is not something I would explain in a getting started guide.

Author: It was for low memory environment (I looked on it because I had some issues with my raspberry).

Reviewer: I think that should be in a separate guide then? 1.21 has [reduce-memory.md](https://github.com/dogecoin/dogecoin/blob/1.21-dev/doc/reduce-memory.md)

Author: I missed it. Do you think I should delete the section entirely, or keep it with a sentence to link it to this file ?

Reviewer: Let's backport it into 1.14.5 (test it to be correct, remove whatever we don't have yet) and then we can link to it here as a "for information on how to reduce the memory footprint, see *link*"?

I'll work on the backport right now.

Reviewer: #2664 adds that document to the 1.14 tree.

Author: Linked.

> **Warning :** For simplicity, explained commands assume executables are under `PATH` environment variable.  
> If needed, specify their location by doing `PATH=$PATH:/path/to/executables` or use their path as command like :
> ```bash
> /path/to/dogecoin-cli [arguments ...]
> ```

### Table of contents

1. [Run a dogecoin node](#run-a-dogecoin-node)
2. [Introduction to JSON-RPC API](#introduction-to-json-rpc-api)
    * [Create a wallet](#create-a-wallet)
    * [Verify balance](#verify-balance)
    * [Send a transaction](#send-a-transaction)
    * [Inspect blocks and transactions](#inspect-blocks-and-transactions)
3. [Node configuration](#node-configuration)
    * [Mainnet, testnet and regnet](#mainnet-testnet-and-regnet)
    * [Data directory](#data-directory)
    * [RPC credentials](#rpc-credentials)
    * [Ports](#ports)
    * [Memory](#memory)
    * [Mode debug](#mode-debug)

## Run a dogecoin node

To start you node, you can run an headless server using `dogecoind` :
```bash
dogecoind -daemon
```

Or you can use a Grapich User Interface (GUI) with `dogecoin-qt` :
```bash
dogecoin-qt
```

Output will be recorded in `debug.log`, located in the [data directory](#data-directory).

> *Use `-help` to see all available options for each executables.*

Reviewer: ```suggestion
*Use `-help` to see all available options for each executable.*
```

Your node should now be running, starting a *synchronization process* by downloading the entire blockchain from other nodes. This operation will take hours to complete... but your now part of the dogecoin network !

> **Warning :** The rest of the guide will assume the use of an headless node. The RPC server is not exposed with `dogecoin-qt` until you activate the `-server` option.

## Introduction to JSON-RPC API

Dogecoin Core expose a JSON-RPC interface. With it, you can request information about blocks and transactions, send transactions to the networks, manage your wallet and so on...

To interact with the interface, it relies on HTTP requests. Using urls like `http://user:password@127.0.0.1:22555/` and sending commands through json format is feasible.

Some tools like libraries can be built on top of it. `dogecoin-cli` is the default shell client provided with Dogecoin Core to use this API.

> The RPC API documentation is available from [bitcoin](https://developer.bitcoin.org/reference/rpc/).  

Author: Moved

To have an overview of all commands, see also :

```bash
# List all commands
dogecoin-cli help

# Get help for a specific command
dogecoin-cli help COMMAND
```

> > **Tips** : Abuse of the help menus is allowed ! They can answer to multiple questions.

Author: Removed. It was to incite people to use the help menu x) They aren't used enough !

### Create a wallet

To receive or recover coins, you will need a wallet address and a private key. The *wallet address* is a shareable information to receive dogecoins from all over the world, while the *private key* is a sensitive information who give the ownership on coins associated to a public address.

By default, you already have a wallet address and a private key available.

List wallet addresses using `getaddressesbyaccount` :

```bash
dogecoin-cli getaddressesbyaccount ""
[
  "DBZc6qeBuo7g8La7L3Lhaf39MsJSgbzpMX"
]
```

Use `getnewaddress` to generate a new wallet address :
```bash
dogecoin-cli getnewaddress

> DC13A8qUgAu5rTWweb44TwXZNQbxDYhyo2

Author: There is a problem about how to display commands, it's why I used the `$`, nothing to render properly a prompt with a smooth copy paste. I already played with different [syntax highlighting](https://docs.github.com/en/github/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks) [source [here](https://github.com/github/linguist/blob/master/lib/linguist/languages.yml)] proposed by Github, sadly they do not seem to not offer something appropriate something appropriate for prompt + output convenient for copy paste.

Don't you think it's better to use `$` for readability over simplicity for copy paste ?  Or there is also console who may look correct:

with `console` keyword:
```console
shibe@doge:~$ echo "command" 43
[
  {
    "txid": "b869ed6606d52e6446dc12db02cf868ab693dd5b9f661116269536f0f8fa2433",
    "vout": 0,
    "address": "nnJDY1xCRgWQc7vBXHUPMPsEynuZW23Y3P",
  }
]
shibetoshi:~$ echo "command" 43
```
With `bash`:
```bash
$ echo "command" 43
[
  {
    "txid": "b869ed6606d52e6446dc12db02cf868ab693dd5b9f661116269536f0f8fa2433",
    "vout": 0,
    "address": "nnJDY1xCRgWQc7vBXHUPMPsEynuZW23Y3P",
  }
]
```

May answer others of your comments where you're suggesting to move code in separate blocks ? Not perfect for copy/pasting but good for readability

For the commented code, it could be :
```console
shibetoshi:~$ dogecoin-cli getnewaddress
DC13A8qUgAu5rTWweb44TwXZNQbxDYhyo2
```
Instead of:
```bash
dogecoin-cli getnewaddress
#return DC13A8qUgAu5rTWweb44TwXZNQbxDYhyo2
```

> DC13A8qUgAu5rTWweb44TwXZNQbxDYhyo2

Reviewer: Agree. The console markup will work well and makes it much clearer when there's a fully decorated prompt (much better than just `$`)

> DC13A8qUgAu5rTWweb44TwXZNQbxDYhyo2

Author: Console actually highlight keyword in comments, look like it is not comments...
```console
# Get help for a specific command
shibetoshi:~$ dogecoin-cli help COMMAND

> DC13A8qUgAu5rTWweb44TwXZNQbxDYhyo2

Author: It's in the same color when comment is fixed to `#`.
```console
#Get help for a specific command
shibetoshi:~$ dogecoin-cli help COMMAND
```

To get details of the 69th block :

```bash
# Find block hash from his height
dogecoin-cli getblockhash 69
3d2def20cd0d3aca148741ef469bda11647a3040d7669c82745d03c728706a8b

# Get block data
dogecoin-cli getblock 3d2def20cd0d3aca148741ef469bda11647a3040d7669c82745d03c728706a8b
{
  "hash": "3d2def20cd0d3aca148741ef469bda11647a3040d7669c82745d03c728706a8b",
  "confirmations": 7816,
  "strippedsize": 190,
  "size": 190,
  "weight": 760,
  "height": 69,
  "version": 1,
  "versionHex": "00000001",
  "merkleroot": "695ce4208fa7a87ef9e99805b0910dc129058ecdceb5cef7e25f71dcdc7936db",
  "tx": [
    "695ce4208fa7a87ef9e99805b0910dc129058ecdceb5cef7e25f71dcdc7936db"
  ],
  "time": 1386475225,
  "mediantime": 1386475209,
  "nonce": 3923708672,
  "bits": "1e0ffff0",
  "difficulty": 0.000244140625,
  "chainwork": "0000000000000000000000000000000000000000000000000000000004600460",
  "previousblockhash": "ffa69e04f928b84f19d84da25fb544340e54dca6c03c33930da245719e61c5ea",
  "nextblockhash": "44bf8abbbb96d4dcfb95df563e606c37987133ea3e013b23bbddde8d7f905fdd"
}
```

See the `tx` field to find the list of all block transactions. Only one exist in the block 69th, a coinbase transaction.

> To get transaction details :

Reviewer: ```suggestion
We can see the entire transaction by querying for its identifier:
```

{
  "hex": "01000000010000000000000000000000000000000000000000000000000000000000000000ffffffff0e04d9eea3520101062f503253482fffffffff0100ac6156be23000023210340a42a5ad6c4c0cd5ae539657032e0a359bd3e0f95771f34d71691b13460a624ac00000000",
  "txid": "695ce4208fa7a87ef9e99805b0910dc129058ecdceb5cef7e25f71dcdc7936db",
  "hash": "695ce4208fa7a87ef9e99805b0910dc129058ecdceb5cef7e25f71dcdc7936db",
  "size": 109,
  "vsize": 109,
  "version": 1,
  "locktime": 0,
  "vin": [
    {
      "coinbase": "04d9eea3520101062f503253482f",
      "sequence": 4294967295
    }
  ],
  "vout": [
    {
      "value": 393004.00000000,
      "n": 0,
      "scriptPubKey": {
        "asm": "0340a42a5ad6c4c0cd5ae539657032e0a359bd3e0f95771f34d71691b13460a624 OP_CHECKSIG",
        "hex": "210340a42a5ad6c4c0cd5ae539657032e0a359bd3e0f95771f34d71691b13460a624ac",
        "reqSigs": 1,
        "type": "pubkey",
        "addresses": [
          "D5TjotnkoXekNQBCn54CBWxjEzCJMMe7WS"
        ]
      }
    }
  ],
  "blockhash": "3d2def20cd0d3aca148741ef469bda11647a3040d7669c82745d03c728706a8b",
  "confirmations": 3964556,
  "time": 1386475225,
  "blocktime": 1386475225

> }

Reviewer: Suggest to put this in its own code block

```

> `vin` and `vout` will give you information about transaction inputs and outputs.

Reviewer: This is confusing because a coinbase transaction does not contain a valid input:

```suggestion
The `vout` structure will give you information about where the transaction output.
```

## Node configuration

> A server contain a lot of parameter to configure, from memory limits, to RPC credentials, ports... There is two ways to prepare your perfect node !
> 
> Using directly `dogecoind` options, `-help` will display all configuration parameters.

Reviewer: ```suggestion
There are many parameters that can be configured to tune your node to your liking. There are two ways to change the configuration.

Using `dogecoind -help` will display all available configuration parameters that can be added as arguments:
```

**Command example :**
```bash
dogecoind -daemon -paytxfee=0.01 -sendfreetransactions=1 -maxconnections=150
```

> Otherwise, a node is configurable through a `dogecoin.conf` file. Create it in the directory defined with the `datadir` setting, `$HOME/.dogecoin` by default, or specify the file location with `-conf`.

Reviewer: ```suggestion
Configuration can be persisted by creating a `dogecoin.conf` file. Create it in the directory defined with the `datadir` setting, `$HOME/.dogecoin` by default, or specify the file location with `-conf`.
```

**dogecoin.conf example :**
```
daemon=1
server=1

> listen=1

Reviewer: As previously discussed, can remove those lines since they are not used in the example above.

Author: I thought we could leave it for GUI users ? Like it, these options exists on the guide.

Reviewer: ok.

paytxfee=0.01
sendfreetransactions=1
maxconnections=150
```
You can see a more concrete example [here](/contrib/debian/examples/dogecoin.conf).

### Mainnet, testnet and regnet

> To perform tests, for example to create a blockchain based application, you may need a sandbox environment. Multiple networks exists for this purpose.

Reviewer: ```suggestion
When trying out new things, for example to test your application that interacts with the Dogecoin chain, it is recommended to not use the main Dogecoin network. Multiple networks are built-in for this purpose.
```

**Mainnet** : The main network where real transaction operate.  
**Testnet** : The test network, with peers.  
**Regnet** : The regression network, to test without peers and create blocks on-demand.

> *Mainnet* is the network used by default. To enable *testnet*, with `dogecoind` use the `-testnet` option and `testnet=1` as `dogecoin.conf` field.

Reviewer: ```suggestion
When not specifying any network, *Mainnet* is the network used by default. To enable *testnet*, use the `dogecoind -testnet`.
```

To enable *regnet*, use `-regtest` option.

> > **Tips :** Think to specify the network when you use `dogecoin-cli`.

Reviewer: ```suggestion
> **Tip:** Remember to specify the network when you want to use `dogecoin-cli`.
```

### Data directory

> The data directory is the location where Dogecoin Core files are stored : `wallet.dat`, `debug.log`, blocks... You can modify it with the `-datadir` setting.

Reviewer: ```suggestion
The data directory is the location where Dogecoin Core files are stored, including the wallet, log files and blocks. You can modify the location with the `-datadir` setting.
```

**Default location :**

Platform | Data directory path
---------|--------------------

> Linux    | `$HOME/.bitcoin/`
> macOS    | `$HOME/Library/Application Support/Bitcoin/`
> Windows  | `%APPDATA%\Bitcoin\`

Reviewer: ```suggestion
Linux    | `$HOME/.dogecoin`
macOS    | `$HOME/Library/Application Support/Dogecoin`
Windows  | `%APPDATA%\Dogecoin`
```

You may need to specify `-datadir` also when using `dogecoin-cli`.

> See [full documentation on file system](/doc/files.md) for more information.

Reviewer: ```suggestion
See the [full documentation on file system](files.md) for more information.
```

### RPC credentials

> A username and a password is needed to control the RPC interface. When no credentials are provided, dogecoin use a [special cookie](https://bitcoin.org/en/release/v0.12.0#rpc-random-cookie-rpc-authentication) to randomly generate it. It's possible to define your own identifiers using `rpcuser` and `rpcpassword` settings.

Reviewer: ```suggestion
Authentication is required to interact with the RPC interface. When no credentials are provided, Dogecoin uses a [random cookie](https://bitcoin.org/en/release/v0.12.0#rpc-random-cookie-rpc-authentication) that gets generated when the software is launched. It's possible to define your own credentials using `rpcuser` and `rpcpassword` parameters.
```

### Ports

> A node exposes up to 2 ports. One mandatory port for the **Peer to Peer Network (P2P)** to communicate with other nodes, a second optional port expose the RPC API.

Reviewer: ```suggestion
A node can expose 2 different ports: one port for the **Peer to Peer Network** (P2P) to communicate with other nodes, and a second port for access to the RPC API. By default, the ports are configured as follows:
```

| Function | mainnet | testnet | regtest |
| :------- | ------: | ------: | ------: |
| P2P      |   22556 |   44556 |   18444 |
| RPC      |   22555 |   44555 |   18332 |

> To configure them use `port` and `rpcport` settings.

Reviewer: ```suggestion
To configure them use the `-port` and `-rpcport` parameters.
```

### Memory

> Dogecoin Core use important ressources for some environment, it may be interesting to restrict them. See [related documentation](/doc/reduce-memory.md) for information on how to reduce the memory footprint.

Reviewer: ```suggestion
Running Dogecoin Core can require a lot of memory, so in some situations it may be necessary to optimize its usage. You can find more information about reducing the memory footprint in the [related guide](reduce-memory.md).
```

### Mode debug

> By default, Dogecoin Core print output to `debug.log`, located in `datadir`. `-printtoconsole` display the server output to the terminal instead of `debug.log`.

Reviewer: ```suggestion
By default, Dogecoin Core keeps detailed logs in the `debug.log` file, located in the `datadir`. Alternatively, the `-printtoconsole` parameter displays the log interactively to the terminal instead.
```

> To get more verbose output with the debug mode, use `-debug` with `dogecoind` or add `debug=1` entry to `dogecoin.conf`.

Reviewer: ```suggestion
To get more verbose log output, you can enable debug mode by using the `-debug=<topic>` parameter to increase logic for a specific topic, or use `-debug=all` to see detailed logs on all topics.
```
